common/src/chain/block/signed_block_header.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crypto::key::Signature;
+use serialization::{Decode, Encode};
+use typename::TypeName;
+
+use crate::primitives::id::{Id, Idable};
+
+use super::{timestamp::BlockTimestamp, Block, BlockHeader, ConsensusData, GenBlock};
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeName)]
+pub enum BlockHeaderSignature {
+    None,
+    PoSBlock(Signature),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeName, serialization::Tagged)]
+pub struct SignedBlockHeader {
+    block_header: BlockHeader,
+    // TODO: Do we want to include the public key to make it possible to cross-check with the spent kernel?
+    signature: BlockHeaderSignature,
+}
+
+impl SignedBlockHeader {
+    pub fn new(signature: BlockHeaderSignature, block_header: BlockHeader) -> Self {
+        Self {
+            signature,
+            block_header,
+        }
+    }
+
+    pub fn signature(&self) -> &BlockHeaderSignature {
+        &self.signature
+    }
+
+    pub fn header(&self) -> &BlockHeader {
+        &self.block_header
+    }
+
+    pub fn header_mut(&mut self) -> &mut BlockHeader {
+        &mut self.block_header
+    }
```

Comment:
I am somewhat uncomfortable with `header_mut` in a signed header. Giving a mutable access to the header changes the header hash which will invalidate any signature that may already be present.

I'd suggest:
* Either document the potential signature integrity violation (possibly even in the method name), or
* remove this method and use `into_block_header`, modify the header, and reconstruct signed header afterwards instead.

Reply:
Another alternative is to return `Option<&mut BlockHeader>` and return a `Some(_)` only if the header signature is `None`.

Reply:
Not a bad idea. I'll go ahead and do it.